src/serac/physics/base_physics.hpp

```diff
@@ -99,13 +93,148 @@ class BasePhysics {
+  {
+    SLIC_ERROR_ROOT_IF(
+        parameter_index >= parameters_.size(),
+        axom::fmt::format("Parameter index '{}' is not available in physics module '{}'", parameter_index, name_));
+
+    SLIC_ERROR_ROOT_IF(
+        !parameters_[parameter_index].state,
+        axom::fmt::format("Parameter index '{}' is not set in physics module '{}'", parameter_index, name_));
+    return *parameters_[parameter_index].state;
+  }
+
+  /**
+   * @brief Get the shape displacement of the associated mesh for this physics object
+   *
+   * @return The associated shape displacement
+   */
+  FiniteElementState& shapeDisplacement() { return shape_displacement_; }
+
+  /// @overload
+  const FiniteElementState& shapeDisplacement() const { return shape_displacement_; }
+
+  /**
+   * @brief Compute the implicit sensitivity of the quantity of interest used in defining the load for the adjoint
+   * problem with respect to the parameter field
+   *
+   * @return The sensitivity with respect to the parameter
+   *
+   * @pre `solveAdjoint` with an appropriate adjoint load must be called prior to this method.
+   */
+  virtual FiniteElementDual& computeSensitivity(size_t /* parameter_index */)
+  {
+    SLIC_ERROR_ROOT(axom::fmt::format("Parameter sensitivities not enabled in physics module {}", name_));
+    return *parameters_[0].sensitivity;
+  }
+
+  /**
+   * @brief Compute the implicit sensitivity of the quantity of interest used in defining the load for the adjoint
+   * problem with respect to the shape displacement field
+   *
+   * @return The sensitivity with respect to the shape displacement
+   *
+   * @pre `solveAdjoint` with an appropriate adjoint load must be called prior to this method.
+   */
+  virtual FiniteElementDual& computeShapeSensitivity()
+  {
+    SLIC_ERROR_ROOT(axom::fmt::format("Shape sensitivities not enabled in physics module {}", name_));
+    return shape_displacement_sensitivity_;
+  }
+
   /**
    * @brief Advance the state variables according to the chosen time integrator
    *
    * @param[inout] dt The timestep to advance. For adaptive time integration methods, the actual timestep is returned.
    */
   virtual void advanceTimestep(double& dt) = 0;
 
+  /**
+   * @brief Solve the adjoint problem
+   * @pre It is expected that the forward analysis is complete and the current state is valid
+   * @note If the essential boundary state is not specified, homogeneous essential boundary conditions are applied
+   *
+   * @return The computed adjoint finite element state
+   */
+  virtual const serac::FiniteElementState& solveAdjoint(FiniteElementDual& /*adjoint_load */,
+                                                        FiniteElementDual* /* dual_with_essential_boundary */ = nullptr)
+  {
+    SLIC_ERROR_ROOT(axom::fmt::format("Adjoint analysis not defined for physics module {}", name_));
+
+    // Return a dummy state value to quiet the compiler. This will never get used.
+    return *states_[0];
+  }
```

Comment:
I think some of these functions should be pure virtual in the future (i.e. once we remove the "Legacy" physics modules), so that the compiler will emit an error message if a new physics module forgets to implement one, instead of defaulting to these incorrect placeholder implementations.